### PR TITLE
#104 Use-schema_translate_map-in-get_lastrowid 

### LINKED ISSUE #104: Use schema_translate_map in get_lastrowid
Hi,

we have an issue when we use autoincrement PKs in combination with `schema_translate_map` set in the engine.
For example:
```python
metadata = MetaData()
my_table = Table(
    "my_table",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("name", String(1000), nullable=False),
    schema="my_schema"
)

engine = create_engine(
        conn_string,
        execution_options={"schema_translate_map": {"my_schema": "my_tenant_schema"}},
    )
conn = engine.connect()
conn.execute(my_table.insert().values(name="foo))
```
throws this error:
```Python Traceback
/caches/pypoetry/virtualenvs/all-KwnZMQER-py3.6/lib/python3.6/site-packages/sqlalchemy/engine/default.py:1273: in _setup_ins_pk_from_lastrowid
    lastrowid = self.get_lastrowid()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <sqlalchemy_exasol.base.EXAExecutionContext object at 0x7f695ec97eb8>

    def get_lastrowid(self):
        columns = self.compiled.sql_compiler.statement.table.columns
        autoinc_pk_columns = \
            [c.name for c in columns if c.autoincrement and c.primary_key]
        if len(autoinc_pk_columns) == 0:
            return None
        elif len(autoinc_pk_columns) > 1:
            util.warn("Table with more than one autoincrement, primary key" \
                      " Column!")
            raise Exception
        else:
            id_col = self.dialect.denormalize_name(autoinc_pk_columns[0])
    
            table = self.compiled.sql_compiler.statement.table.name
            table = self.dialect.denormalize_name(table)
    
            sql_stmnt = "SELECT column_identity from SYS.EXA_ALL_COLUMNS " \
                        "WHERE column_object_type = 'TABLE' and column_table " \
                        "= ? AND column_name = ?"
    
            schema = self.compiled.sql_compiler.statement.table.schema
            if schema is not None:
                schema = self.dialect.denormalize_name(schema)
                sql_stmnt += " AND column_schema = ?"
            cursor = self.create_cursor()
            if schema:
                cursor.execute(sql_stmnt, (table, id_col, schema))
            else:
                cursor.execute(sql_stmnt, (table, id_col))
            result = cursor.fetchone()
            cursor.close()
>           return int(result[0]) - 1
E           TypeError: 'NoneType' object is not subscriptable

/caches/pypoetry/virtualenvs/all-KwnZMQER-py3.6/lib/python3.6/site-packages/sqlalchemy_exasol/base.py:337: TypeError
```
in version 2.1.

For our particular case, the error can be circumvented by using the schema in the `schema_translate_map` in the error throwing `get_lastrowid` function:
```python
            if schema is not None:
                schema = self.compiled.sql_compiler.schema_translate_map.map_[schema] if schema in self.compiled.sql_compiler.schema_translate_map.map_ else schema # <--- explicitely use the translation map here 
                schema = self.dialect.denormalize_name(schema)
                sql_stmnt += " AND column_schema = ?"
```

### PR DESCRIPTION
**Pull request by Nicoretti**
For futher details see https://github.com/exasol/sqlalchemy-exasol/issues/104

### CODE REVIEW

File: sqlalchemy_exasol/base.py [hunk 1]
```diff
@@ -307,36 +307,42 @@ def get_insert_default(self, column):
...
+            raise Exception(msg)
+
+        id_col = self.dialect.denormalize_name(autoinc_pk_columns[0])
+        table = self.compiled.sql_compiler.statement.table.name
+        table = self.dialect.denormalize_name(table)
+
+        def _schema(sql_compiler, dialect):
+            """Get the schema while taking the translation-map and the de-normalization into account"""
+            translate_map = sql_compiler.schema_translate_map
+            schema_dispatcher = dict(translate_map.map_) if translate_map else {}
```
Comment (tkilias) [inline sqlalchemy_exasol/base.py]: I though the map_ is already a dict

File: sqlalchemy_exasol/base.py [hunk 2]
```diff
@@ -307,36 +307,42 @@ def get_insert_default(self, column):
...
+        if len(autoinc_pk_columns) > 1:
+            msg = "Table with more than one autoincrement, primary key Column!"
+            util.warn(msg)
+            raise Exception(msg)
+
+        id_col = self.dialect.denormalize_name(autoinc_pk_columns[0])
+        table = self.compiled.sql_compiler.statement.table.name
+        table = self.dialect.denormalize_name(table)
+
+        def _schema(sql_compiler, dialect):
```
Comment (tkilias) [inline sqlalchemy_exasol/base.py]:
```suggestion
        def _get_schema(sql_compiler, dialect):
```

File: sqlalchemy_exasol/base.py [hunk 3]
```diff
@@ -307,36 +307,42 @@ def get_insert_default(self, column):
...
+        schema = _schema(
+            self.compiled.sql_compiler,
+            self.dialect
+        )
+
+        sql_stmnt = " ".join((
+            "SELECT column_identity from SYS.EXA_ALL_COLUMNS",
+            "WHERE column_object_type = 'TABLE' and column_table",
+            "= ? AND column_name = ?",
+            "AND column_schema = ?" if schema else ""
```
Comment (tkilias) [inline sqlalchemy_exasol/base.py]: a bit to much magic for my taste, please split into to commands, or put parts into variables

**Review by tkilias — CHANGES_REQUESTED** [review]

File: sqlalchemy_exasol/base.py (hunk 1 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/base.py]: good point

**Review by tkilias — APPROVED** [review]

**Pull request merged**